### PR TITLE
redis functionality added, message handling retry when call ongoing

### PR DESCRIPTION
**Pull request by thanaParis**
redis cache now being used.

- redis docker image rebuilt to include custom redis.conf for bind 0.0.0.0 and protected-mode no
- RedisCache now accepts config to create client, logs connection events, connects to client

message handling retry when call ongoing
- queue receivers (IMessageHandler implementations) listen for custom error RetryMessageError and do not ack message if it is thrown
- RetryMessageError is thrown when a call is received by the central system for a station with an ongoing call
- N.B.: Calls are ongoing until: a CallResult is received, a CallError is received, timeout occurs
  - websocket message timeout is in config, websocket.maxCallLengthSeconds

### CODE REVIEW

File: Server/src/assets/logo.png.license [hunk 1]
```diff
@@ -0,0 +1,3 @@
+Copyright Contributors to the CitrineOS Project
```
Comment (ChrisWeissmann) [inline Server/src/assets/logo.png.license:1]: What's the correct license now? on all the other files I saw:
```
// Copyright (c) 2023 S44, LLC
// Copyright Contributors to the CitrineOS Project
//
// SPDX-License-Identifier: Apache 2.0
```

File: 00_Base/src/interfaces/modules/AbstractModule.ts [hunk 2]
```diff
@@ -164,7 +153,7 @@ export abstract class AbstractModule implements IModule {
     }
 
     async handleMessageApiCallback(message: IMessage<OcppResponse>): Promise<void> {
-        const url: string | null = await this._cache.getAndRemove(message.context.correlationId, this.CALLBACK_URL_CACHE_PREFIX + message.context.stationId);
+        const url: string | null = await this._cache.get(message.context.correlationId, this.CALLBACK_URL_CACHE_PREFIX + message.context.stationId);
```
[the remark this answers is not in the document]
Comment (thanaParis) [reply 00_Base/src/interfaces/modules/AbstractModule.ts:156]: it's removed when it times out instead. I think this is safer. I found that the redis 'get and remove' was not actually atomic, and there wasn't an easy way to make it so, unfortunately.

File: 02_Util/src/cache/memory.ts [hunk 3]
```diff
@@ -25,7 +14,6 @@ export class MemoryCache implements ICache {
   private _cache: Map<string, string>;
   private _keySubscriptionMap: Map<string, (arg: string | null) => void>;
   private _keySubscriptionPromiseMap: Map<string, Promise<string | null>>;
-  private _lockMap: Map<string, Promise<void>>;
```
[the remark this answers is not in the document]
Comment (thanaParis) [reply 02_Util/src/cache/memory.ts:28]: it was only needed for getAndRemove, which was removed

File: 02_Util/src/cache/redis.ts [hunk 4]
```diff
@@ -102,31 +94,21 @@ export class RedisCache implements ICache {
...
+      return false;
     });
   }
 
-  set(key: string, value: string, namespace?: string, expireSeconds?: number): Promise<boolean> {
+  setIfNotExist(key: string, value: string, namespace?: string, expireSeconds?: number): Promise<boolean> {
     namespace = namespace || "default";
     key = `${namespace}:${key}`;
-    return this._client.set(key, value, { EX: expireSeconds }).then((result) => {
+    return this._client.set(key, value, { EX: expireSeconds, NX: true }).then((result) => {
```
[the remark this answers is not in the document]
Comment (thanaParis) [reply 02_Util/src/cache/redis.ts:111]: excellent point. I don't think there's any calls to set that don't include an expiration time, which is how this went unnoticed--this causes the method to fail at runtime if expireSeconds is undefined. We should discuss whether expiration time being optional makes sense to begin with. using the cache without an expiration, or with a default expiration, are both dangerous...

File: Server/src/assets/logo.png.license (hunk 1 above)
Comment (thanaParis) [reply Server/src/assets/logo.png.license:1]: the copyright from S44 is legacy, it shouldn't be removed but won't be used going forward. the SPDX header missing here is weird... the reuse tool made this file automatically when adding the headers. I'll need to check whether the tool is bugged or if there's a good reason the spdx header is missing.

**Pull request merged**